site/js/SessionActor.js

```diff
@@ -101,48 +131,72 @@ export default class SessionActor extends Actor {
   }
 
   async promptRegister() {
-    const username = prompt('Username?')
-
-    if (username === null) {
-      this.emit('registration cancel')
-      return
-    }
-
-    const password = prompt(
-      'Password? (Must be at least 6 characters long.)\n' +
-      'This password is NOT secure yet! It is stored securely, but you are probably not on an HTTPS connection.\n' +
-      'That means just about anyone can look at your HTTP request and find out your password.\n' +
-      'DO NOT use something you use anywhere else!'
-    )
-
-    if (password === null) {
-      this.emit('registration cancel')
-      return
+    let username, password
+
+    try {
+      username = await this.actors.modals.prompt(
+        'Register', 'Username?', '',
+        async name => {
+          const reValid = /^[a-zA-Z0-9-_]+$/
+
+          if (name.length === 0) {
+            throw 'Please enter a username.'
+          } else if (!reValid.test(name)) {
+            throw 'Usernames cannot contain special characters other than - and _.'
+          }
+
+          // TODO check if username is taken or not here, in advance
+        },
+        'Continue', 'Cancel')
+    } catch(error) {
+      if (error === 'modal closed') {
+        this.emit('registration cancel', 1)
+
+        return
+      } else {
+        throw error
+      }
```

Comment:
I still wish there were built-in syntax for "catch this particular thing, otherwise throw". Alas!

```js
try {
  ...
} catch (('modal closed')) {
  this.emit('registration cancel', 1)
  return
}
// but nope, that's not real syntax!
```